RFC for replica IDs and tombstones.

bdarnell opened the pull request.

bdarnell committed:
RFC for replica IDs and tombstones.

tschottdorf commented on docs/RFCS/replica_tombstone.md:
```diff
@@ -0,0 +1,101 @@
+
+We have an additional complication compared to vanilla raft because we
+allow node IDs to be reused (this is necessary for coalesced
+heartbeats. We may remove a replica from a node and then re-add it
```
s/heartbeats/heartbeats)/

tschottdorf commented on docs/RFCS/replica_tombstone.md:
```diff
@@ -0,0 +1,101 @@
+# Unresolved questions
+
+* How long must we wait before garbage-collecting tombstones? Can we
+  do it at all?
```
we can always do it when the corresponding Range a) isn't initialized b) is being garbage collected, or can we not?

bdarnell replied:
A) Uninitialized ranges are just in-memory placeholders; they don't have any data on disk until they are initialized by a snapshot. So they don't need to leave a tombstone and don't even really need to be GC'd. 

B) I don't understand this point. Tombstones are written by the GC process so by definition the replica they refer to has been GC'd.

tschottdorf replied:
sorry, the comment here was clearly written with my brain turned off.

tschottdorf commented on docs/RFCS/replica_tombstone.md:
```diff
@@ -0,0 +1,101 @@
+  do it at all?
+* How will the raft transport get the correct replica ID? Will
+  `MultiRaft` itself have to be aware of replica IDs for proper
+  synchronization?
```
If the `To:` `uint64` encoded the replica ID, that would get us a long way (I know, we're already mushing up `NodeID` and `RangeID` in it). Conceptually, `ReplicaID` is the unique identifier of a consensus group member and storing it in Raft makes sense. Maybe that's also something we could try to get into upstream, though I don't see why that would be a worthwhile addition for their use case. In any case, if we had that, Raft would have all the information needed when sending, and the receiving end can (in a convenient location) peruse the `ReplicaID`. Refusing stale messages early always needs to hit storage somehow, but sure seems possible to fire out a one-off RPC to tell a stale group's MultiRaft instance that its stale incarnation can go so that sustained traffic from a stale source is extremely unlikely (also appears less central than the core issues addressed in this RFC).

bdarnell replied:
Hmm, that's a possibility. Raft only really needs to know about the ReplicaID, and in fact if we give it replica IDs instead of store IDs we wouldn't even need tombstones. We'd just need `MultiRaft` to be able to map group + replica IDs to node + store IDs to route messages and coalesce heartbeats. I think a leader would always be able to query the `Store` to get the current members of the range and perform this mapping. The tricky thing would be routing responses for followers. We'd need to remember the source address of any request with a previously-unseen `From` (or be able to use our RPC connections bidirectionally). This at least deserves to be written up as an alternative.

tschottdorf commented:
I think this is a good solution. Finding a nice and clean implementation will require some thought, but conceptually this looks sound (and doesn't require much added effort).

vivekmenezes commented on docs/RFCS/replica_tombstone.md:
```diff
@@ -0,0 +1,101 @@
+   range R, it starts an election.
+5. Nodes A and B see that node C has a more advanced log position for
+   range R than they do (since they have nothing), so they vote for it.
+   C becomes leader and sends snapshots to A and B.
```
Can we invent a mechanism to guarantee that A and B have a more advanced log position, say by inserting a dummy write that increments the index of the range. When A and B garbage collect their own replicas they also delete their sense of the (term,index) for the range right?

bdarnell replied:
A and B don't have a log position because they've garbage collected their log. Currently, they can't tell whether C is the legitimate leader of the range adding them as a replica or if it's out of date. I originally considered using the raft (term, index) in the tombstone instead of introducing replica IDs, but I'm not sure we can guarantee safety with this approach. The problem is that even if we insert a dummy write, there is no guarantee that that write will reach nodes A and B.

bdarnell committed:
Replica IDs replace RaftNodeIDs.

This answers the question of how MultiRaft knows about the right replica
ID, and also avoids some additional edge cases.

bdarnell commented:
I've updated the RFC based on an idea from @tschottdorf: we replace the current constructed `RaftNodeID` with the `ReplicaID`, which closes off a few more edge cases.

bdarnell committed:
Mark RFC as in-progress

spencerkimball commented on docs/RFCS/replica_tombstone.md:
```diff
@@ -0,0 +1,113 @@
+
+# Unresolved questions
+
+* How long must we wait before garbage-collecting tombstones? Can we
```
I'm not sure this question needs to be resolved. Let's not cache the tombstones. It should be rare to get a message for which there's no legitimate range. Either we're creating a new one in the normal course of rebalancing or repair, or it's an old message--both should be exceptions compared to the normal flow of traffic. Since we're able to go to disk, the total number of tombstones can be incredibly large without causing much of a headache.

spencerkimball commented:
LGTM
